### PR TITLE
[feat] - implement AddressSet CRD and support in firewall controller

### PR DESCRIPTION
**Pull request by AshleyDumaine**
**What this PR does / why we need it**: Adds a new CRD that can be reused to construct LinodeFirewalls in case addresses are common between firewalls (e.g. common allowed hosts for control plane and worker node Firewalls)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**TODOs**:

- [x] squashed commits
- [x] includes documentation
- [x] adds unit tests
- [x] adds or updates e2e tests

### CODE REVIEW

File: api/v1alpha2/linodefirewall_types.go [hunk 1]
```diff
@@ -70,6 +70,9 @@ type FirewallRule struct {
 	// +kubebuilder:validation:Enum=TCP;UDP;ICMP;IPENCAP
 	Protocol  linodego.NetworkProtocol `json:"protocol"`
 	Addresses *NetworkAddresses        `json:"addresses"`
+	// AddressSetRefs is a list of references to AddressSets
+	// If Addresses is present, AddressSetRefs will be ignored
+	AddressSetRefs []*corev1.ObjectReference `json:"addressSetRefs,omitempty"`
```
Comment (eljohnson92) [inline api/v1alpha2/linodefirewall_types.go:75]: do these have to be mutually exclusive or could we do both?

Comment (AshleyDumaine) [reply api/v1alpha2/linodefirewall_types.go:75]: We could do both if it's less confusing

File: internal/controller/linodefirewall_controller_helpers.go [hunk 2]
```diff
@@ -162,34 +234,62 @@ func transformToCIDR(ip string) string {
...
 // processInboundRule handles a single inbound rule
-func processInboundRule(rule infrav1alpha2.FirewallRule, createOpts *linodego.FirewallCreateOptions) {
-	ruleIPv4s, ruleIPv6s := processAddresses(rule.Addresses)
+func processInboundRule(ctx context.Context, k8sClient clients.K8sClient, log logr.Logger, rule infrav1alpha2.FirewallRule, lfw *infrav1alpha2.LinodeFirewall, createOpts *linodego.FirewallCreateOptions) error {
+	var ruleIPv4s []string
+	var ruleIPv6s []string
+	var err error
+	if rule.Addresses != nil {
+		ruleIPv4s, ruleIPv6s = processAddresses(rule.Addresses)
+	} else if rule.AddressSetRefs != nil {
```
Comment (tchinmai7) [inline internal/controller/linodefirewall_controller_helpers.go:243]: should this be elseIf ? if the intent is to support both?

File: internal/controller/linodefirewall_controller_helpers.go [hunk 3]
```diff
@@ -199,20 +299,77 @@ func processAddresses(addresses *infrav1alpha2.NetworkAddresses) (ipv4s, ipv6s [
...
+	for _, addrSetRef := range addressSetRefs {
+		addrSet := &infrav1alpha2.AddressSet{}
+		if addrSetRef.Namespace == "" {
+			addrSetRef.Namespace = lfw.Namespace
+		}
+		if err := k8sClient.Get(ctx, client.ObjectKey{Namespace: addrSetRef.Namespace, Name: addrSetRef.Name}, addrSet); err != nil {
+			log.Error(err, "failed to fetch referenced AddressSet", "namespace", addrSetRef.Namespace, "name", addrSetRef.Name)
+			return nil, nil, err
+		}
+		finalizer := fmt.Sprintf("lfw.%s.%s", lfw.Namespace, lfw.Name)
```
Comment (tchinmai7) [inline internal/controller/linodefirewall_controller_helpers.go:341]: should we add a helper to the firewalltype? `getFinalizer()`?

File: internal/controller/linodefirewall_controller_helpers.go (hunk 2 above)
Comment (AshleyDumaine) [reply internal/controller/linodefirewall_controller_helpers.go:248]: looks like that was leftover, outbound was updated but not inbound

File: internal/controller/linodefirewall_controller_helpers.go [hunk 4]
```diff
@@ -26,15 +35,81 @@ var (
...
+}
+
+// Constructs a unique list of requests for updating LinodeFirewalls that either reference the
+// AddressSet in the Inbound and/or OutboundRules
+func buildRequests(firewalls []infrav1alpha2.LinodeFirewall, obj client.Object) []reconcile.Request {
+	requestSet := make(map[reconcile.Request]struct{})
+	for _, item := range firewalls {
+		for _, inboundRule := range item.Spec.InboundRules {
+			for _, addrSetRef := range inboundRule.AddressSetRefs {
+				if addrSetRef.Name == obj.GetName() && addrSetRef.Namespace == obj.GetNamespace() {
```
Comment (geoffreylefebvre) [inline internal/controller/linodefirewall_controller_helpers.go:70]: Could the inner loop section be moved into its own function, along with the section in the loop below?

File: internal/controller/linodefirewall_controller.go [hunk 5]
```diff
@@ -203,6 +205,32 @@ func (r *LinodeFirewallReconciler) reconcile(
...
+		addrSetRef.Namespace = fwScope.LinodeFirewall.Namespace
+	}
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: addrSetRef.Namespace, Name: addrSetRef.Name}, addrSet); err != nil {
+		if !apierrors.IsNotFound(err) {
+			logger.Error(err, "failed to fetch referenced AddressSet", "namespace", addrSetRef.Namespace, "name", addrSetRef.Name)
+			return err
+		}
+	}
+	finalizer := fmt.Sprintf("lfw.%s.%s", fwScope.LinodeFirewall.Namespace, fwScope.LinodeFirewall.Name)
+	if controllerutil.ContainsFinalizer(addrSet, finalizer) {
```
Comment (tchinmai7) [inline internal/controller/linodefirewall_controller.go:223]: i don't think we need to check before calling `removeFinalizer` - we can replace this as
```
if controllerutil.RemoveFinalizer()... {
}
```

File: internal/controller/linodefirewall_controller.go [hunk 6]
```diff
@@ -250,6 +293,9 @@ func (r *LinodeFirewallReconciler) SetupWithManager(mgr ctrl.Manager, options cr
 	if err != nil {
 		return fmt.Errorf("failed to create mapper for LinodeFirewalls: %w", err)
 	}
+	if err != nil {
+		return fmt.Errorf("failed to create mapper for AddressSets: %w", err)
+	}
```
Comment (tchinmai7) [inline internal/controller/linodefirewall_controller.go:298]: what error is this referring to?

File: internal/controller/linodefirewall_controller_helpers.go [hunk 7]
```diff
@@ -26,15 +35,81 @@ var (
...
+						NamespacedName: types.NamespacedName{
+							Name:      item.GetName(),
+							Namespace: item.GetNamespace(),
+						},
+					}] = struct{}{}
+				}
+			}
+		}
+	}
+	requests := make([]reconcile.Request, 0, len(requestSet))
```
Comment (geoffreylefebvre) [inline internal/controller/linodefirewall_controller_helpers.go:93]: I think go 1.23 has maps.Values that returns an iterator. You can combine it with slices.Collect() to get a slice.

**Comment by rahulait** [comment]
The PR looks good to me. However, I do see a race condition where it doesn't properly update the firewall rules. For example, I have these two configs:
File 1.yaml
```
---
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
kind: AddressSet
metadata:
  name: test-addr-set
spec:
  ipv4:
  - "10.0.0.0/8"
  - "192.168.0.0/16"
  ipv6:
  - ::/0
```

File 2.yaml
```
---
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
kind: LinodeFirewall
metadata:
  name: kube4
  namespace: default
spec:
  enabled: false
  inboundPolicy: DROP
  inboundRules:
  - action: ACCEPT
    addresses:
      ipv4:
      - 10.0.0.0/8
    description: accept all tcp traffic within the vpc
    label: intra-cluster-tcp
    ports: 1-65535
    protocol: TCP
  - action: ACCEPT
    label: test-addr-set
    ports: "1-65535"
    protocol: "TCP"
    addressSetRefs:
    - name: test-addr-set
      kind: AddressSet
---
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
kind: AddressSet
metadata:
  name: test-addr-set
spec:
  ipv4:
  - "10.0.0.0/8"
  ipv6:
  - ::/0
```

I am doing following operations:
```
kubectl apply -f 1.yaml
kubectl apply -f 2.yaml
```

I am expecting that when I apply 2.yaml, it should create a firewall on linode with rules and those rules should contain only one ipv4 address (10.0.0.0/8). However, sometimes I do see multiple addresses (10.0.0.0/8 and 192.168.0.0/16). I do see addressset getting updated when 2.yaml is applied, but firewall uses the older addressset values. CAPL doesn't fix it unless addressset is modified or capl-controller-manager is restarted. It doesn't happen 100% of the time but does happen when retried 1-2 times. This points me to some kind of race condition.

File: internal/controller/linodefirewall_controller.go (hunk 6 above)
Comment (AshleyDumaine) [reply internal/controller/linodefirewall_controller.go:298]: leftovers I forgot to remove

File: internal/controller/linodefirewall_controller_helpers.go [hunk 8]
```diff
@@ -64,37 +66,31 @@ func findObjectsForAddressSet(logger logr.Logger, tracedClient client.Client) ha
...
-					requestSet[reconcile.Request{
-						NamespacedName: types.NamespacedName{
-							Name:      item.GetName(),
-							Namespace: item.GetNamespace(),
-						},
-					}] = struct{}{}
-				}
-			}
+		for _, outboundRule := range firewall.Spec.OutboundRules {
+			requestSet = buildRequestsHelper(requestSet, firewall, outboundRule.AddressSetRefs, obj)
```
Comment (geoffreylefebvre) [inline internal/controller/linodefirewall_controller_helpers.go:74]: Aren't you squashing the inbound rule content of requestSet?

**Comment by AshleyDumaine** [comment]
> The PR looks good to me. However, I do see a race condition where it doesn't properly update the firewall rules. For example, I have these two configs: File 1.yaml
> 
> ```
> ---
> apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
> kind: AddressSet
> metadata:
>   name: test-addr-set
> spec:
>   ipv4:
>   - "10.0.0.0/8"
>   - "192.168.0.0/16"
>   ipv6:
>   - ::/0
> ```
> 
> File 2.yaml
> 
> ```
> ---
> apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
> kind: LinodeFirewall
> metadata:
>   name: kube4
>   namespace: default
> spec:
>   enabled: false
>   inboundPolicy: DROP
>   inboundRules:
>   - action: ACCEPT
>     addresses:
>       ipv4:
>       - 10.0.0.0/8
>     description: accept all tcp traffic within the vpc
>     label: intra-cluster-tcp
>     ports: 1-65535
>     protocol: TCP
>   - action: ACCEPT
>     label: test-addr-set
>     ports: "1-65535"
>     protocol: "TCP"
>     addressSetRefs:
>     - name: test-addr-set
>       kind: AddressSet
> ---
> apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
> kind: AddressSet
> metadata:
>   name: test-addr-set
> spec:
>   ipv4:
>   - "10.0.0.0/8"
>   ipv6:
>   - ::/0
> ```
> 
> I am doing following operations:
> 
> ```
> kubectl apply -f 1.yaml
> kubectl apply -f 2.yaml
> ```
> 
> I am expecting that when I apply 2.yaml, it should create a firewall on linode with rules and those rules should contain only one ipv4 address (10.0.0.0/8). However, sometimes I do see multiple addresses (10.0.0.0/8 and 192.168.0.0/16). I do see addressset getting updated when 2.yaml is applied, but firewall uses the older addressset values. CAPL doesn't fix it unless addressset is modified or capl-controller-manager is restarted. It doesn't happen 100% of the time but does happen when retried 1-2 times. This points me to some kind of race condition.

I tried this 10 times but wasn't able to reproduce the issue with the latest changes 🤔

**Comment by rahulait** [comment]
> I tried this 10 times but wasn't able to reproduce the issue with the latest changes 🤔

Are you checking it on linode UI? This is what I see with latest branch as well:
Screenshot 2024-12-04 at 11 59 14 AM

However, on k8s, its:
```
(devbox) rahsharm@my-linux-vm:~/cluster-api-provider-linode$ k apply -f 1.yaml
addressset.infrastructure.cluster.x-k8s.io/test-addr-set created
(devbox) rahsharm@my-linux-vm:~/cluster-api-provider-linode$ k apply -f 2.yaml
linodefirewall.infrastructure.cluster.x-k8s.io/kube4 created
addressset.infrastructure.cluster.x-k8s.io/test-addr-set configured
(devbox) rahsharm@my-linux-vm:~/cluster-api-provider-linode$ k get addressset test-addr-set -o yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
kind: AddressSet
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"infrastructure.cluster.x-k8s.io/v1alpha2","kind":"AddressSet","metadata":{"annotations":{},"name":"test-addr-set","namespace":"default"},"spec":{"ipv4":["10.0.0.0/8"],"ipv6":["::/0"]}}
  creationTimestamp: "2024-12-04T16:57:25Z"
  finalizers:
  - lfw.default.kube4
  generation: 2
  name: test-addr-set
  namespace: default
  resourceVersion: "301373"
  uid: 93dc465b-4436-40c6-ab62-bccc01700752
spec:
  ipv4:
  - 10.0.0.0/8
  ipv6:
  - ::/0
(devbox) rahsharm@my-linux-vm:~/cluster-api-provider-linode$
```

**Review by rahulait — APPROVED** [review]
LGTM. The issue is not easily reproducible and seems to be happening only on my mgmt cluster running on ubuntu. We'll keep an eye on this and see if we can narrow it down easily in future.

**Pull request merged**